crates/gpui/src/platform/windows/window.rs

```diff
@@ -0,0 +1,536 @@
+    }
+
+    // todo!("windows")
+    fn on_moved(&self, callback: Box<dyn FnMut()>) {
+        self.inner.callbacks.borrow_mut().moved = Some(callback);
+    }
+
+    // todo!("windows")
+    fn on_should_close(&self, callback: Box<dyn FnMut() -> bool>) {
+        self.inner.callbacks.borrow_mut().should_close = Some(callback);
+    }
+
+    // todo!("windows")
+    fn on_close(&self, callback: Box<dyn FnOnce()>) {
+        self.inner.callbacks.borrow_mut().close = Some(callback);
+    }
+
+    // todo!("windows")
+    fn on_appearance_changed(&self, callback: Box<dyn FnMut()>) {
+        self.inner.callbacks.borrow_mut().appearance_changed = Some(callback);
+    }
+
+    // todo!("windows")
+    fn is_topmost_for_position(&self, position: Point<Pixels>) -> bool {
+        true
+    }
+
+    // todo!("windows")
+    fn draw(&self, scene: &Scene) {
+        self.inner.renderer.borrow_mut().draw(scene)
+    }
+
+    // todo!("windows")
+    fn sprite_atlas(&self) -> Arc<dyn PlatformAtlas> {
+        self.inner.renderer.borrow().sprite_atlas().clone()
+    }
+}
+
+fn register_wnd_class() -> PCWSTR {
+    const CLASS_NAME: PCWSTR = w!("Zed::Window");
+
+    static ONCE: Once = Once::new();
+    ONCE.call_once(|| {
+        let wc = WNDCLASSW {
+            lpfnWndProc: Some(wnd_proc),
+            hCursor: unsafe { LoadCursorW(None, IDC_ARROW).ok().unwrap() },
+            lpszClassName: PCWSTR(CLASS_NAME.as_ptr()),
+            ..Default::default()
+        };
+        unsafe { RegisterClassW(&wc) };
+    });
+
+    CLASS_NAME
+}
+
+unsafe extern "system" fn wnd_proc(
+    hwnd: HWND,
+    msg: u32,
+    wparam: WPARAM,
+    lparam: LPARAM,
+) -> LRESULT {
+    if msg == WM_NCCREATE {
+        let cs = lparam.0 as *const CREATESTRUCTW;
+        let cs = unsafe { &*cs };
+        let ctx = cs.lpCreateParams as *mut WindowCreateContext;
+        let ctx = unsafe { &mut *ctx };
+        let inner = Rc::new(WindowsWindowInner::new(
+            hwnd,
+            cs,
+            ctx.platform_inner.clone(),
+            ctx.handle,
+        ));
+        let weak = Box::new(Rc::downgrade(&inner));
+        unsafe { set_window_long(hwnd, GWLP_USERDATA, Box::into_raw(weak) as isize) };
+        ctx.inner = Some(inner);
+        return LRESULT(1);
+    }
+    let ptr = unsafe { get_window_long(hwnd, GWLP_USERDATA) } as *mut Weak<WindowsWindowInner>;
+    if ptr.is_null() {
+        return unsafe { DefWindowProcW(hwnd, msg, wparam, lparam) };
+    }
+    let inner = unsafe { &*ptr };
+    let r = if let Some(inner) = inner.upgrade() {
+        inner.handle_msg(msg, wparam, lparam)
+    } else {
+        unsafe { DefWindowProcW(hwnd, msg, wparam, lparam) }
+    };
+    if msg == WM_NCDESTROY {
+        unsafe { set_window_long(hwnd, GWLP_USERDATA, 0) };
+        unsafe { std::mem::drop(Box::from_raw(ptr)) };
+    }
+    r
+}
+
+unsafe fn get_window_long(hwnd: HWND, nindex: WINDOW_LONG_PTR_INDEX) -> isize {
+    #[cfg(target_pointer_width = "64")]
```

Comment:
do we really need to support 32-bit windows systems?

Reply:
Well, Windows 10 32bit x86 ISOs are still available for download, `i686-pc-windows-msvc` is a Tier 1 target for Rust.
But if we look at VSCode it doesn't provide binaries for that, also not all language server binaries may support it.

Reply:
https://github.com/zed-industries/zed/pull/8490#discussion_r1504559362
This change is what I made in response to this suggestion